src/main/kotlin/org/rust/lang/core/resolve/RustResolveEngine.kt

```diff
@@ -91,32 +134,29 @@ object RustResolveEngine {
             val pathPart = path.pathPart ?: return
 
             val isPlainPathImport = path.mul == null && path.lbrace == null
+
+            // `use foo::bar as baz;`
             if (isPlainPathImport) {
                 val name = path.alias ?: pathPart
                 if (match(name)) {
-                    if (use in visited) {
+                    if (use in visitedImports) {
                         return
                     }
-                    visited += use
+                    visitedImports += use
 
-                    val item = RustResolveEngine.resolve(pathPart, visited).element  ?: return
+                    val item = resolve(pathPart).element  ?: return
                     return found(item)
                 }
             }
-        }
-
 
-        private fun seek(elem: RustDeclaringElement) {
-            seek(listOf(elem))
         }
 
+        private fun seek(elem: RustDeclaringElement) = seek(listOf(elem))
+
         private fun seek(decls: Collection<RustDeclaringElement>) {
             decls.flatMap { it.getBoundElements() }
-                 .forEach { e ->
-                     if (match(e)) {
-                         return found(e)
-                     }
-                 }
+                 .find{match(it)}
+                 ?.let {found(it)}
```

Comment:
It is a bit alien precisely because it ends a chain of collections operations and invokes an operation on the single element. 

I don't see a way to improve it much: we can make `found` to accept `RustNamedElement?` but it would be only worse. 

We can revert back to imperative `forEach` version, but it is more verbose and I find this `find/found` rhyme neat :)